File: Fourth Semester - Design and Programming Environments/Java/ProiectJavaNetworking/TravelModel/src/java/org/example/model/validator/ValidatorFactory.java
package org.example.model.validator;


public class ValidatorFactory implements Factory {
    private static ValidatorFactory instance = null;

    private ValidatorFactory() {
    }

    public static ValidatorFactory getInstance() {
        if (instance == null) {
            instance = new ValidatorFactory();
        }
        return instance;
    }

    @Override
    public Validator createValidator(ValidatorStrategy strategy) {
        switch (strategy) {
            case TOUR:
                return new TourValidator();
            case RESERVATION:
                return new ReservationValidator();
            case AGENT:
                return new AgentValidator();
            default:
                return null;
        }
    }
}

File: Fourth Semester - Design and Programming Environments/C#/ProiectC#Networking/TravelModel/model/validator/ValidatorFactory.cs
namespace TravelModel.model.validator;

public class ValidatorFactory
{
    private static ValidatorFactory _instance;
    
    private ValidatorFactory() { }

    public static ValidatorFactory GetInstance()
    {
        return _instance ??= new ValidatorFactory();
    }

    public IValidator<T> CreateValidator<T>(ValidatorStrategy strategy)
    {
        return strategy switch
        {
            ValidatorStrategy.Agent => new AgentValidator() as IValidator<T>,
            ValidatorStrategy.Reservation => new ReservationValidator() as IValidator<T>,
            ValidatorStrategy.Tour => new TourValidator() as IValidator<T>,
            _ => throw new NotImplementedException()
        };
    }
}